smart_open/s3.py

```diff
@@ -890,22 +936,42 @@ def tell(self):
     def detach(self):
         raise io.UnsupportedOperation("detach() not supported")
 
-    def write(self, b):
+    def write(self, b: Buffer) -> int:
         """Write the given buffer (bytes, bytearray, memoryview or any buffer
         interface implementation) to the S3 file.
 
         For more information about buffers, see https://docs.python.org/3/c-api/buffer.html
 
         There's buffering happening under the covers, so this may not actually
         do any HTTP transfer right away."""
+        #
+        # Part size: 5 MiB to 5 GiB. There is no minimum size limit on the last
+        # part of your multipart upload.
 
-        length = self._buf.write(b)
-        self._total_bytes += length
+        # botocore does not accept memoryview, otherwise we could've gotten
+        # away with not needing to write a copy to the buffer aside from cases
+        # where b is smaller than min_part_size
+        #
 
-        if self._buf.tell() >= self._min_part_size:
-            self._upload_next_part()
+        offset = 0
+        mv = memoryview(b)
+        self._total_bytes += len(mv)
 
-        return length
+        while offset < len(mv):
+            start = offset
+            end = offset + self._max_part_size - self._buf.tell()
```

Comment:
@jakkdl Do we need to keep separate min and max part size variables in MultipartWriter?

I think a single variable is enough.  We can continue calling it min_part_size in the public API, but it really just means "part size", so "once you have this many bytes, upload a new part".

WDYT?

Reply:
I think it could be quite useful for end users, either if they have flaky internet and would prefer sending multiple small chunks over a big 5GiB transfer that might fail halfway through, or if they're sending a big file on disk and want to limit memory usage. (rip no memoryview)

Reply:
In that case they can set the min_part_size to something smaller, e.g. 5MB.

What's the benefit of having a separate max_part_size parameter?

Reply:
I'm going to get rid of it for now, because it's an entirely new parameter.

If people end up asking for it at some point in the future, we can discuss adding it.

Reply:
Right, I guess it's only if they're sending their file object in one go (or in very big chunks) that max_part_size would make a difference - and they could write a pretty simple wrapper themselves that chunks it into multiple smaller calls.